File: app/src/main/java/com/example/sleppify/ExoMediaPlayer.java
package com.example.sleppify;

import android.content.Context;
import android.media.AudioAttributes;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.media3.common.C;
import androidx.media3.common.MediaItem;
import androidx.media3.common.MimeTypes;
import androidx.media3.common.PlaybackException;
import androidx.media3.common.Player;
import androidx.media3.common.util.UnstableApi;
import androidx.media3.datasource.DataSource;
import androidx.media3.datasource.DefaultDataSource;
import androidx.media3.datasource.okhttp.OkHttpDataSource;
import androidx.media3.exoplayer.ExoPlayer;
import androidx.media3.exoplayer.source.DefaultMediaSourceFactory;
import androidx.media3.exoplayer.source.ProgressiveMediaSource;
import androidx.media3.exoplayer.hls.HlsMediaSource;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;

/**
 * Thin adapter around ExoPlayer (Media3) that mimics the relevant subset of
 * {@link android.media.MediaPlayer} API used by {@link SongPlayerFragment}.
 *
 * <p>The goal is to swap MediaPlayer implementations with minimal changes to
 * the fragment. Only the listeners and APIs actually used are exposed.
 */
@UnstableApi
public class ExoMediaPlayer {

    public interface OnPreparedListener {
        void onPrepared(@NonNull ExoMediaPlayer mp);
    }

    public interface OnCompletionListener {
        void onCompletion(@NonNull ExoMediaPlayer mp);
    }

    public interface OnErrorListener {
        /**
         * @return true if the error has been handled
         */
        boolean onError(@NonNull ExoMediaPlayer mp, int what, int extra);
    }

    private static final String TAG = "ExoMediaPlayer";

    private final Context appContext;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private ExoPlayer exoPlayer;

    @Nullable private OnPreparedListener preparedListener;
    @Nullable private OnCompletionListener completionListener;
    @Nullable private OnErrorListener errorListener;

    @Nullable private Uri pendingUri;
    @Nullable private String pendingPath;
    @Nullable private Map<String, String> pendingHeaders;
    private boolean pendingIsHttpSource;
    private boolean prepared;
    private boolean released;
    private int audioSessionId;
    private float leftVolume = 1f;
    private float rightVolume = 1f;

    public ExoMediaPlayer(@NonNull Context context) {
        this.appContext = context.getApplicationContext();
        this.exoPlayer = new ExoPlayer.Builder(appContext).build();
        this.audioSessionId = this.exoPlayer.getAudioSessionId();
        this.exoPlayer.addListener(playerListener);
    }

    private final Player.Listener playerListener = new Player.Listener() {
        @Override
        public void onPlaybackStateChanged(int playbackState) {
            if (playbackState == Player.STATE_READY && !prepared) {
                prepared = true;
                if (preparedListener != null) {
                    mainHandler.post(() -> {
                        if (!released && preparedListener != null) {
                            preparedListener.onPrepared(ExoMediaPlayer.this);
                        }
                    });
                }
            } else if (playbackState == Player.STATE_ENDED) {
                if (completionListener != null) {
                    mainHandler.post(() -> {
                        if (!released && completionListener != null) {
                            completionListener.onCompletion(ExoMediaPlayer.this);
                        }
                    });
                }
            }
        }

        @Override
        public void onPlayerError(@NonNull PlaybackException error) {
            Log.w(TAG, "onPlayerError: code=" + error.errorCode + " message=" + error.getMessage());
            if (errorListener != null) {
                mainHandler.post(() -> {
                    if (!released && errorListener != null) {
                        errorListener.onError(ExoMediaPlayer.this, error.errorCode, 0);
                    }
                });
            }
        }
    };

    public void setAudioAttributes(@NonNull AudioAttributes attrs) {
        if (released || exoPlayer == null) return;
        // Map framework AudioAttributes to Media3
        androidx.media3.common.AudioAttributes media3Attrs = new androidx.media3.common.AudioAttributes.Builder()
                .setUsage(C.USAGE_MEDIA)
                .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
                .build();
        exoPlayer.setAudioAttributes(media3Attrs, false);
    }

    public void setOnPreparedListener(@Nullable OnPreparedListener l) {
        this.preparedListener = l;
    }

    public void setOnCompletionListener(@Nullable OnCompletionListener l) {
        this.completionListener = l;
    }

    public void setOnErrorListener(@Nullable OnErrorListener l) {
        this.errorListener = l;
    }

    public void setDataSource(@NonNull Context context, @NonNull Uri uri, @Nullable Map<String, String> headers) {
        this.pendingUri = uri;
        this.pendingPath = null;
        this.pendingHeaders = headers != null ? new HashMap<>(headers) : null;
        String scheme = uri.getScheme();
        this.pendingIsHttpSource = "http".equalsIgnoreCase(scheme) || "https".equalsIgnoreCase(scheme);
    }

    public void setDataSource(@NonNull String path) {
        this.pendingPath = path;
        this.pendingUri = null;
        this.pendingHeaders = null;
        this.pendingIsHttpSource = path.startsWith("http://") || path.startsWith("https://");
    }

    /**
     * Alias of {@link #prepareAsync()} — ExoPlayer always prepares asynchronously.
     * Callers that relied on MediaPlayer's blocking {@code prepare()} should
     * wait for {@link OnPreparedListener} before interacting.
     */
    public void prepare() {
        prepareAsync();
    }

    public void prepareAsync() {
        if (released || exoPlayer == null) {
            throw new IllegalStateException("ExoMediaPlayer released");
        }
        Uri uri = pendingUri != null ? pendingUri : (pendingPath != null ? Uri.parse(pendingPath) : null);
        if (uri == null) {
            throw new IllegalStateException("No data source set");
        }

        DataSource.Factory factory;
        if (pendingIsHttpSource) {
            OkHttpClient okHttpClient = new OkHttpClient.Builder()
                    .connectTimeout(15, TimeUnit.SECONDS)
                    .readTimeout(15, TimeUnit.SECONDS)
                    .followRedirects(true)
                    .followSslRedirects(true)
                    .build();

            OkHttpDataSource.Factory httpFactory = new OkHttpDataSource.Factory(okHttpClient);

            if (pendingHeaders != null && !pendingHeaders.isEmpty()) {
                httpFactory.setDefaultRequestProperties(pendingHeaders);
            }

            factory = httpFactory;
        } else {
            factory = new DefaultDataSource.Factory(appContext);
        }

        MediaItem mediaItem = MediaItem.fromUri(uri);
        androidx.media3.exoplayer.source.MediaSource source = new ProgressiveMediaSource.Factory(factory)
                .createMediaSource(mediaItem);

        exoPlayer.setMediaSource(source);
        exoPlayer.setVolume(Math.max(leftVolume, rightVolume));
        exoPlayer.prepare();
        prepared = false;
    }

    public void start() {
        if (released || exoPlayer == null) return;
        exoPlayer.setPlayWhenReady(true);
        if (exoPlayer.getPlaybackState() == Player.STATE_IDLE) {
            exoPlayer.prepare();
        }
    }

    public void pause() {
        if (released || exoPlayer == null) return;
        exoPlayer.setPlayWhenReady(false);
    }

    public void stop() {
        if (released || exoPlayer == null) return;
        exoPlayer.stop();
    }

    public boolean isPlaying() {
        if (released || exoPlayer == null) return false;
        return exoPlayer.isPlaying()
                || (exoPlayer.getPlayWhenReady()
                    && exoPlayer.getPlaybackState() != Player.STATE_IDLE
                    && exoPlayer.getPlaybackState() != Player.STATE_ENDED);
    }

    public void seekTo(int msec) {
        if (released || exoPlayer == null) return;
        exoPlayer.seekTo(Math.max(0L, msec));
    }

    public int getCurrentPosition() {
        if (released || exoPlayer == null) return 0;
        return (int) Math.max(0L, exoPlayer.getCurrentPosition());
    }

    public int getDuration() {
        if (released || exoPlayer == null) return 0;
        long d = exoPlayer.getDuration();
        if (d == C.TIME_UNSET || d < 0) return 0;
        return (int) d;
    }

    public void setVolume(float left, float right) {
        this.leftVolume = left;
        this.rightVolume = right;
        if (released || exoPlayer == null) return;
        exoPlayer.setVolume(Math.max(left, right));
    }

    public int getAudioSessionId() {
        if (released || exoPlayer == null) return audioSessionId;
        return exoPlayer.getAudioSessionId();
    }

    public void release() {
        if (released) return;
        released = true;
        prepared = false;
        preparedListener = null;
        completionListener = null;
        errorListener = null;
        if (exoPlayer != null) {
            try {
                exoPlayer.removeListener(playerListener);
                exoPlayer.stop();
                exoPlayer.release();
            } catch (Exception e) {
                Log.w(TAG, "release: exception", e);
            }
            exoPlayer = null;
        }
    }
}

File: app/src/main/java/com/example/sleppify/ExoMediaPlayer.kt
package com.example.sleppify

import android.content.Context
import android.media.AudioAttributes
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.common.util.UnstableApi
import androidx.media3.datasource.DataSource
import androidx.media3.datasource.DefaultDataSource
import androidx.media3.datasource.DefaultHttpDataSource
import androidx.media3.datasource.cache.CacheDataSource
import androidx.media3.datasource.cache.LeastRecentlyUsedCacheEvictor
import androidx.media3.datasource.cache.SimpleCache
import androidx.media3.exoplayer.DefaultLoadControl
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.exoplayer.source.ProgressiveMediaSource
import java.io.File
import java.util.concurrent.CopyOnWriteArrayList
import java.lang.ref.WeakReference

/**
 * Thin adapter around ExoPlayer (Media3) that mimics the relevant subset of
 * [android.media.MediaPlayer] API used by [SongPlayerFragment].
 *
 * The goal is to swap MediaPlayer implementations with minimal changes to
 * the fragment. Only the listeners and APIs actually used are exposed.
 */
@UnstableApi
class ExoMediaPlayer {

    interface OnPreparedListener {
        fun onPrepared(mp: ExoMediaPlayer)
    }

    interface OnCompletionListener {
        fun onCompletion(mp: ExoMediaPlayer)
    }

    interface OnErrorListener {
        /**
         * @return true if the error has been handled
         */
        fun onError(mp: ExoMediaPlayer, what: Int, extra: Int): Boolean
    }

    companion object {
        private const val TAG = "ExoMediaPlayer"
        private const val CACHE_SIZE_BYTES = 100L * 1024 * 1024 // 100 MB disk cache

        @Volatile
        private var sharedCache: SimpleCache? = null
        private val cacheLock = Any()

        @JvmStatic
        fun getSharedCache(context: Context): SimpleCache {
            return sharedCache ?: synchronized(cacheLock) {
                sharedCache ?: run {
                    val cacheDir = File(context.cacheDir, "exo_stream_cache")
                    val evictor = LeastRecentlyUsedCacheEvictor(CACHE_SIZE_BYTES)
                    SimpleCache(cacheDir, evictor, androidx.media3.database.StandaloneDatabaseProvider(context)).also {
                        sharedCache = it
                    }
                }
            }
        }

        // Registry of all active ExoMediaPlayer instances in the app
        private val activeInstances = CopyOnWriteArrayList<WeakReference<ExoMediaPlayer>>()

        /**
         * Pauses all other playing instances that are not part of a crossfade.
         */
        @JvmStatic
        private fun stopOthers(current: ExoMediaPlayer) {
            val iterator = activeInstances.iterator()
            while (iterator.hasNext()) {
                val ref = iterator.next()
                val other = ref.get()
                if (other == null) {
                    activeInstances.remove(ref)
                    continue
                }
                if (other !== current && other.isPlaying()) {
                    // Allow overlap if one of them is a crossfade component
                    if (!current.isCrossfadeComponent && !other.isCrossfadeComponent) {
                        Log.d(TAG, "stopOthers: pausing competing player instance")
                        try {
                            other.pause()
                        } catch (e: Exception) {
                            Log.w(TAG, "Failed to pause other instance", e)
                        }
                    }
                }
            }
        }
    }

    private val appContext: Context
    private val mainHandler = Handler(Looper.getMainLooper())
    private var exoPlayer: ExoPlayer? = null

    private var preparedListener: OnPreparedListener? = null
    private var completionListener: OnCompletionListener? = null
    private var errorListener: OnErrorListener? = null

    private var pendingUri: Uri? = null
    private var pendingPath: String? = null
    private var pendingHeaders: Map<String, String>? = null
    private var pendingIsHttpSource: Boolean = false
    private var prepared: Boolean = false
    private var released: Boolean = false
    private var ownsPlayer: Boolean = false
    private var audioSessionId: Int = 0
    private var leftVolume = 1f
    private var rightVolume = 1f
    
    /**
     * Mark this instance as part of a crossfade transition.
     * Crossfade components are allowed to overlap with other players briefly.
     */
    @JvmField
    var isCrossfadeComponent: Boolean = false

    constructor(context: Context) {
        this.appContext = context.applicationContext
        val audioAttributes = androidx.media3.common.AudioAttributes.Builder()
            .setUsage(C.USAGE_MEDIA)
            .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
            .build()

        val loadControl = DefaultLoadControl.Builder()
            .setBufferDurationsMs(15_000, 50_000, 500, 1_000)
            .setPrioritizeTimeOverSizeThresholds(true)
            .build()

        val player = ExoPlayer.Builder(appContext)
            .setLoadControl(loadControl)
            .setAudioAttributes(audioAttributes, true)
            .setHandleAudioBecomingNoisy(true)
            .build()
        this.exoPlayer = player
        this.ownsPlayer = true
        this.audioSessionId = player.audioSessionId
        player.addListener(playerListener)
        activeInstances.add(WeakReference(this))
    }

    /**
     * Constructor that uses a shared ExoPlayer instance to reduce startup latency.
     * @param context Application context
     * @param sharedExoPlayer Pre-initialized ExoPlayer instance
     */
    constructor(context: Context, sharedExoPlayer: ExoPlayer) {
        this.appContext = context.applicationContext
        this.exoPlayer = sharedExoPlayer
        this.ownsPlayer = false
        this.audioSessionId = sharedExoPlayer.audioSessionId
        sharedExoPlayer.addListener(playerListener)
        activeInstances.add(WeakReference(this))
    }

    private val playerListener = object : Player.Listener {
        override fun onPlaybackStateChanged(playbackState: Int) {
            if (playbackState == Player.STATE_READY && !prepared) {
                prepared = true
                preparedListener?.let { listener ->
                    mainHandler.post {
                        if (!released) {
                            listener.onPrepared(this@ExoMediaPlayer)
                        }
                    }
                }
            } else if (playbackState == Player.STATE_ENDED) {
                completionListener?.let { listener ->
                    mainHandler.post {
                        if (!released) {
                            listener.onCompletion(this@ExoMediaPlayer)
                        }
                    }
                }
            }
        }

        override fun onPlayerError(error: PlaybackException) {
            Log.w(TAG, "onPlayerError: code=${error.errorCode} message=${error.message}")
            errorListener?.let { listener ->
                mainHandler.post {
                    if (!released) {
                        listener.onError(this@ExoMediaPlayer, error.errorCode, 0)
                    }
                }
            }
        }
    }

    fun setAudioAttributes(attrs: AudioAttributes) {
        if (released) return
        val player = exoPlayer ?: return
        // Map framework AudioAttributes to Media3
        val media3Attrs = androidx.media3.common.AudioAttributes.Builder()
            .setUsage(C.USAGE_MEDIA)
            .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
            .build()
        player.setAudioAttributes(media3Attrs, false)
    }

    fun setOnPreparedListener(l: OnPreparedListener?) {
        this.preparedListener = l
    }

    fun setOnCompletionListener(l: OnCompletionListener?) {
        this.completionListener = l
    }

    fun setOnErrorListener(l: OnErrorListener?) {
        this.errorListener = l
    }

    fun setDataSource(context: Context, uri: Uri, headers: Map<String, String>?) {
        this.pendingUri = uri
        this.pendingPath = null
        this.pendingHeaders = headers?.let { HashMap(it) }
        val scheme = uri.scheme
        this.pendingIsHttpSource = "http".equals(scheme, ignoreCase = true) || "https".equals(scheme, ignoreCase = true)
    }

    fun setDataSource(path: String) {
        this.pendingPath = path
        this.pendingUri = null
        this.pendingHeaders = null
        this.pendingIsHttpSource = path.startsWith("http://") || path.startsWith("https://")
    }

    /**
     * Alias of [prepareAsync] — ExoPlayer always prepares asynchronously.
     * Callers that relied on MediaPlayer's blocking `prepare()` should
     * wait for [OnPreparedListener] before interacting.
     */
    fun prepare() {
        prepareAsync()
    }

    fun prepareAsync() {
        if (released) {
            throw IllegalStateException("ExoMediaPlayer released")
        }
        val player = exoPlayer ?: throw IllegalStateException("ExoMediaPlayer released")
        val uri = pendingUri ?: pendingPath?.let { Uri.parse(it) } ?: throw IllegalStateException("No data source set")

        val factory: DataSource.Factory = if (pendingIsHttpSource) {
            // Usar DefaultHttpDataSource (HttpURLConnection nativo de Android)
            // en vez de OkHttpDataSource — el TLS fingerprint de OkHttp es detectado
            // por el CDN de YouTube causando 403, mientras que HttpURLConnection
            // tiene un fingerprint consistente con el de una app Android real.
            val httpFactory = DefaultHttpDataSource.Factory()
                .setConnectTimeoutMs(8_000)
                .setReadTimeoutMs(8_000)
                .setAllowCrossProtocolRedirects(true)

            pendingHeaders?.let { headers ->
                if (headers.isNotEmpty()) {
                    httpFactory.setDefaultRequestProperties(headers)
                    Log.d(TAG, "prepareAsync: setting ${headers.size} headers on DefaultHttpDataSource: ${headers.keys}")
                }
            }
            // Wrap with CacheDataSource for disk caching of streamed audio
            CacheDataSource.Factory()
                .setCache(getSharedCache(appContext))
                .setUpstreamDataSourceFactory(httpFactory)
                .setFlags(CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR)
        } else {
            DefaultDataSource.Factory(appContext)
        }

        val mediaItem = MediaItem.fromUri(uri)
        val source = ProgressiveMediaSource.Factory(factory)
            .createMediaSource(mediaItem)

        player.setMediaSource(source)
        player.volume = Math.max(leftVolume, rightVolume)
        player.prepare()
        prepared = false
    }

    fun start() {
        if (released) return
        val player = exoPlayer ?: return
        
        // Enforce single-playback policy (Global Guardian)
        stopOthers(this)
        
        player.playWhenReady = true
        if (player.playbackState == Player.STATE_IDLE) {
            player.prepare()
        }
    }

    fun pause() {
        if (released) return
        exoPlayer?.playWhenReady = false
    }

    fun stop() {
        if (released) return
        exoPlayer?.stop()
    }

    fun isPlaying(): Boolean {
        if (released) return false
        val player = exoPlayer ?: return false
        return player.isPlaying || (player.playWhenReady &&
                player.playbackState != Player.STATE_IDLE &&
                player.playbackState != Player.STATE_ENDED)
    }

    fun seekTo(msec: Int) {
        if (released) return
        exoPlayer?.seekTo(Math.max(0L, msec.toLong()))
    }

    fun getCurrentPosition(): Int {
        if (released) return 0
        return (exoPlayer?.currentPosition?.coerceAtLeast(0L)?.toInt()) ?: 0
    }

    fun getDuration(): Int {
        if (released) return 0
        val d = exoPlayer?.duration ?: 0L
        if (d == C.TIME_UNSET || d < 0) return 0
        return d.toInt()
    }

    fun setVolume(left: Float, right: Float) {
        this.leftVolume = left
        this.rightVolume = right
        if (released) return
        exoPlayer?.volume = Math.max(left, right)
    }

    fun getAudioSessionId(): Int {
        if (released) return audioSessionId
        return exoPlayer?.audioSessionId ?: audioSessionId
    }

    fun release() {
        if (released) return
        released = true
        prepared = false
        preparedListener = null
        completionListener = null
        errorListener = null
        exoPlayer?.let { player ->
            try {
                player.removeListener(playerListener)
                if (ownsPlayer) {
                    player.stop()
                    player.clearMediaItems()
                    player.release()
                }
            } catch (e: Exception) {
                Log.w(TAG, "release: exception", e)
            }
            exoPlayer = null
        }
    }
}
